M2/Macaulay2/packages/HomologicalShiftIdeals.m2

```diff
@@ -0,0 +1,1001 @@
+        else for i from 0 to n do (
+            j=l#i;
+            u=u*S_i^j;
+        );
+        u
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the multidegree of a given monomial
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    toMultidegree
+    (toMultidegree,RingElement)
+  Headline
+    compute the multidegree of a monomial
+  Usage
+    toMultidegree u
+  Inputs
+    u:RingElement
+      a monomial $u=x_1^{a_1}\cdots x_n^{a_n}$
+  Outputs
+    l:List
+      the multidegree $(a_1,\dots,a_n)$ of $u$
+  Description
+    Text
+      The multidegree of a monomial $u\in S=K[x_1,\ldots,x_n]$ is the vector $(a_1,a_2,\dots,a_n)$ where $a_i=\max\{j:x_i^j\ \text{divides}\ u\}$.
+    Example
+      S = QQ[x_1..x_4];
+      u = x_2^2*x_3*x_4^2;
+      toMultidegree(u)
+///);
+
+toMultidegree = method(TypicalValue=>List)
+toMultidegree (RingElement) := u -> (
+	S:=ring u;
+        if not isMonomialIdeal(ideal(u)) then error
+        else n:=numgens S-1;
+        L:={};
+        for i from 0 to n do (
+              mdeg:={};
+              for j from 0 to n do (
+                     if j==i then mdeg=append(mdeg,1)
+                     else mdeg=append(mdeg,0);
+              );
+              L=append(L,mdeg);
+        );
+        R:=newRing(S,Degrees=>L);
+        f:=map(R,S);
+        degree f u
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the bounding multidegree of a monomial ideal
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    boundingMultidegree
+    (boundingMultidegree,Ideal)
+  Headline
+    compute the bounding multidegree of a monomial ideal
+  Usage
+    boundingMultidegree I
+  Inputs
+    I:Ideal
+      a monomial ideal $I\subset S=K[x_1,\dots,x_n]$, $K$ a field
+  Outputs
+    l:List
+      the bounding multidegree of $I$: $\textbf{deg}(I)$
+  Description
+    Text
+      The bounding multidegree of a monomial ideal $I$ of a polynomial ring $S=K[x_1,\ldots,x_n]$ is the vector $\textbf{deg}(I)=(\text{deg}_{x_1}(I),\text{deg}_{x_2}(I),\dots,\text{deg}_{x_n}(I))$ with $$\text{deg}_{x_i}(I)=\max_{u\in G(I)}\max\{j:x_i^j\ \text{divides}\ u\},$$ where $G(I)$ is the minimal monomial generating set of $I$.
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      boundingMultidegree(I)
+///);
+
+boundingMultidegree = method(TypicalValue=>List)
+boundingMultidegree (Ideal) := I -> (
+	if not isMonomialIdeal(I) then error
+        else S:=ring I;
+        n:=numgens S-1;
+        L:=apply(flatten entries mingens I,x->toMultidegree(x));
+        bDeg:={};
+        for i from 0 to n do (
+              bDeg=append(bDeg,max(apply(L,x->x#i)));
+        );
```

Comment:
Again, repeatedly appending inside a for loop isn't great.  Here's a quick way to do the same thing:

```m2
max \ transpose L
```